vcd/resource_vcd_vapp_vm.go

```diff
@@ -1719,17 +1821,68 @@ func resourceVcdVAppVmUpdateExecute(d *schema.ResourceData, meta interface{}, ex
 			if d.HasChange("os_type") {
 				vmSpecSection.OsType = d.Get("os_type").(string)
 			}
-
 			if d.HasChange("description") {
 				description = d.Get("description").(string)
 			}
+			if d.HasChange("firmware") {
+				firmware := d.Get("firmware").(string)
+
+				hw, err := vdc.GetHardwareVersion(d.Get("hardware_version").(string))
+				if err != nil {
+					return diag.Errorf("error getting hardware version: %s", err)
+				}
+
+				os, err := vdc.FindOsFromId(hw, d.Get("os_type").(string))
+				if err != nil {
+					return diag.Errorf("error finding given OS type in the provided hardware version: %s", err)
+				}
+
+				if !contains(os.SupportedFirmware, firmware) {
+					return diag.Errorf("the OS type doesn't support provided firmware")
+				}
 
-			_, err := vm.UpdateVmSpecSection(vmSpecSection, description)
+				// Additional check needs to be made in a scenario where EFI secure boot is
+				// disabled and firmware is set to 'bios' during update, as there would be a misconfiguration error
+				// because EFI secure boot can't be enabled when firmware is not efi and we can't do two things at once
+
+				efiSecureBoot := d.Get("boot_options.0.efi_secure_boot").(bool)
+				if firmware != "efi" && efiSecureBoot {
+					return diag.Errorf("error: EFI secure boot can only be enabled with firmware set to 'efi'")
+				}
+				if firmware != "efi" && !efiSecureBoot {
+					efiSecureBootOptions := &types.BootOptions{
+						EfiSecureBootEnabled: addrOf(efiSecureBoot),
+					}
+					vm, err = vm.UpdateBootOptions(efiSecureBootOptions)
```

Comment:
It's a good point, although I'm not completely sure what's the best method around it. Would it be better to remove the `Refresh()` call completely or just not update the `vm` variable? 

There can be a risk of overriding other configuration, but then it's probably the other code that needs fixing because we shouldn't override the `vm` variable ourselves, only with API calls

Reply:
It is somewhat safer to have an explicit `Refresh` call (easier to see when reading). There is also this comment as part of refactoring that might be useful https://github.com/Didainius/terraform-provider-vcd/blob/edge-uplinks/vcd/resource_vcd_vapp_vm.go#L53-L63